### PR TITLE
Implemented filtering by checked out

### PR DESCRIPTION
**Pull request by nvyush**
Pull Request for Issue #16959 .

### Summary of Changes
Added filtering by checked out

### Testing Instructions
Install patch, view lists of:
- banners;
- banner categories;
- banners clients;
- contacts;
- contacts categories;
- articles;
- featured articles;
- articles categories;
- menu items of site;
- menu items of admin panel;
- modules of site;
- modules of admin panel;
- newsfeeds;
- newsfeeds categories;
- plugins;
- tags;
- user notes;
- user notes categories;

Select search tools to show more search options and a new filter is on the right. called "- Select Checked out -".

### CODE REVIEW

**Comment by wojsmol** [comment]
@nvyush Please use `__DEPLOY_VERSION__` in `@since` tags instead of 3.8.

**Comment by nvyush** [comment]
@wojsmol Is OK now?

**Comment by Quy** [comment]
Tested with 3.8 staging. Error under Contents > Articles:

`filter[checked_out] has invalid value of the sql_from attribute. `

File: administrator/language/en-GB/en-GB.ini [hunk 1]
```diff
@@ -881,6 +881,7 @@ JOPTION_SELECT_PUBLISHED="- Select Status -"
 JOPTION_SELECT_TEMPLATE="- Select Template -"
 JOPTION_SELECT_MAX_LEVELS="- Select Max Levels -"
 JOPTION_SELECT_TAG="- Select Tag -"
+JOPTION_SELECT_CHECKED_OUT="- Select Checked out -"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.ini]: Capitalize `Out`.

**Comment by bayareajenn** [comment]
Greetings. Tested with 3.8 beta and once applied if one goes to Content -> Articles we get:

An error has occurred.

    0 filter[checked_out] has invalid value of the sql_from attribute. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17491">issues.joomla.org/tracker/joomla-cms/17491</a>.</sub>

**Comment by crleathers** [comment]
I have tested this item :red_circle: unsuccessfully on 8868a8f581180ce2ae7792c81bf866c57cff1884<br /><br />In the backend, the Article Manager gives this error with the patch installed:
 filter[checked_out] has invalid value of the sql_from attribute.

Uninstalling the patch clears the error.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17491">issues.joomla.org/tracker/joomla-cms/17491</a>.</sub>

**Comment by nvyush** [comment]
Oops, at the first variant ```sql_from``` had default value "#__content c", later I decided to do it required, but xml for articles leaved as was.
It must be OK now.

**Comment by bayareajenn** [comment]
I have tested this item :white_check_mark: successfully on 8868a8f581180ce2ae7792c81bf866c57cff1884<br /><br />The issue with the article manager error is now resolved. Tested the filtering by Checked Out. Works well.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17491">issues.joomla.org/tracker/joomla-cms/17491</a>.</sub>

**Comment by nvyush** [comment]
@bayareajenn thanks for testing. I believe you made a mistake with the commit number, the last patched commit is ceb166d.

**Comment by bayareajenn** [comment]
@nvyush really? I just went to this issue number in patch tester (17491) and applied the patch and it worked. Forgive me, I'm not the best at this testing thing but I'm getting better. 

Is applying the patch from patch tester not sufficient? Or is there a different patch/issue number I should test? Cuz what I did gave me a successful test. Let me know and forgive delays, I'm in California and behind almost the entire rest of the world.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17491">issues.joomla.org/tracker/joomla-cms/17491</a>.</sub>

**Comment by nvyush** [comment]
@bayareajenn On github.com [this PR](https://github.com/joomla/joomla-cms/pull/17491) has last commit of ceb166d. May be this has not matter.

**Comment by bayareajenn** [comment]
No, it's the same and I did test and it was successful. It matters not whatever that six digit number is. The patch tester is pulling the new PR for us to test. (I think I said that right.)

So as for me, your new PR works and I tested successfully.

**Comment by mbabker** [comment]
FWIW the commit thing is done by the issue tracker, has nothing to do with patch tester.  Could just mean the data for this issue fell out of sync on the tracker.  Either way it's generally not a big deal.

**Comment by bayareajenn** [comment]
@mbabker it's worth a lot. The more I learn the more I understand making me a more valuable tester that needs less lame guidance. So thank you very much. 

Regardless, I tested the fix and it works as does the filtering by checked out now.

**Comment by nvyush** [comment]
@mbabker Thanks a lot. I thought that only approved commit will be merged.
@crleathers Could you recheck last patch?

**Comment by infograf768** [comment]
I suggest a small improvement.

add in getOptions()
```
			foreach ($options as $option)
			{
				$option->text = \JText::sprintf('JBY', $option->text);
			}
```
and in en-GB.ini
`JBY="by %s"`

to obtain
screen shot 2017-08-16 at 11 09 20

What do you think?

**Comment by brianteeman** [comment]
it would read better without the "by Super User" and instead changed the
- Select Checked Out -
to
- Select Checked Out By -

**Comment by infograf768** [comment]
Agree

**Comment by nvyush** [comment]
It was corrected as suggested by brianteeman. Please, recheck.

**Comment by brianteeman** [comment]
Why is there a "none" in the select? I dont understand that option<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17491">issues.joomla.org/tracker/joomla-cms/17491</a>.</sub>

**Comment by brianteeman** [comment]
Having tested this PR I really dont see the value

**Comment by bayareajenn** [comment]
I was thinking something similar to @brianteeman earlier. Here's how my thought process went:

Check-in from the article manager can be handy indeed. But the only reason I see to have this filter there is so that a Super Admin or Administrator doesn't have to go to System -> Global Check in.

I then thought, well maybe if there's some UG/ACL that doesn't allow someone to get to System. 

But, if a user that is in the Manager User Group tries to check in an Article that is checked out by a Super, they can't check it in anyways. And they shouldn't be able to.

When I train people, one of the things I tell them is that if they need to get into an article that is locked, the first thing they need to do is check to see if anyone is logged in before they "check-in" an article. That way they don't blow someone's work away by checking in an article if someone else is working on it.  

This filter might be handy to save 5 clicks for a Super, but not really. And for those that can't check in anyways due to permissions, it's just going to frustrate them that they can't do anything to check in anyways.

**Comment by bayareajenn** [comment]
PS I didn't realise I could question something that was to test. I was just "testing." Am I allowed to do that?

**Comment by brianteeman** [comment]
@bayareajenn of course you can. We dont merge code just because it works ;)

**Comment by bayareajenn** [comment]
Thanks, @brianteeman I appreciate that. I'll allow myself to express myself in the future.

**Comment by nvyush** [comment]
"None" selects all items, that are not "checked out".
I add it because it can be useful to someone and it is effortless to implement. If the community believes this choice is harmful, I'll delete it.

**Comment by brianteeman** [comment]
Well none is the same  them as having nothing selected?

**Comment by bayareajenn** [comment]
I'm not saying it's "harmful", I just don't know that there's a good reason for it.

**Comment by nvyush** [comment]
@brianteeman No, if there are "checked out" items, then empty filter selects all items, but "None" selects all "not checked out" items.
About of the value of the filter. If we have many authors that work simultaneously and we need check-in some articles, we cannot use System -> Global Check in otherwise we drop work all users. In my case the same for contacts.

**Comment by bayareajenn** [comment]
Can you train them to not use the "back" button in the browser when they finish working on a contact or article? That would keep them from staying checked out in the first place? @nvyush

**Comment by nvyush** [comment]
@bayareajenn All users are trained and know how to end the work with articles and contacts, but sometimes this happens.

**Comment by bayareajenn** [comment]
@nvyush I know it does. I'm just trying to sort out why you would need more than the individual items where you check the box and then click the "Check-in" button from the Top Toolbar in those cases. If it happens so often that isn't sufficient, then maybe a refreshing training is in order. ;)

**Comment by mbabker** [comment]
For me just reading this conversation, the All/None filters aren't necessarily bad, but they could be better labeled (All Users and Not Checked Out as an example).

For practica use, I can't say I have a case where the None filter adds value, but I also don't do enough work with users/clients who aren't trained in working with Joomla where I'd see a value in either adding it or excluding it from the filter.  So I guess I'm not picky either way if it exists.

**Comment by bayareajenn** [comment]
I tested the commit 10c5731 and see the new label "Select Checked Out By" fine.

**Comment by nvyush** [comment]
@bayareajenn If you have more than 10000 articles and more than 1000 contacts, it is difficult to find needed items, isn't it?
@mbabker I used the available translations and found them quite informative. I did not want to overload the translation file with additional translations. Maybe I was wrong.

**Comment by bayareajenn** [comment]
I totally get it. But not all of them are checked out (I hope). If someone goes to one they need into and it is checked out, then they can check that one in (if no one else is logged in and editing it). But all good. I did test and it does work. :)

**Comment by brianteeman** [comment]
You do know there is an extension available that will check everything in after a set time?

**Comment by infograf768** [comment]
> For me just reading this conversation, the All/None filters aren't necessarily bad, but they could be better labeled (All Users and Not Checked Out as an example).

Imho, `All Users` if fine.
But as we have now `Select Checked Out By`, `Not Checked Out` reads a bit weird.

**Comment by nvyush** [comment]
@brianteeman We have the command button "Check In" in Joomla! already without any extensions. Using of it without filtering by "checked out" field is inconvenient. We either need to add filtering, or remove the button. I suggest to add filtering.
 @infograf768 My English is not very good to offer acceptable translations. Tell me, please, how to fix it and I will do it. If "None" is superfluous, let's just take it away.

**Comment by bayareajenn** [comment]
I vote for "None" being removed.

**Comment by nvyush** [comment]
I replace "All" with "All Users" and remove "None". Is it OK now?

File: administrator/language/en-GB/en-GB.ini [hunk 2]
```diff
@@ -881,6 +882,7 @@ JOPTION_SELECT_PUBLISHED="- Select Status -"
 JOPTION_SELECT_TEMPLATE="- Select Template -"
 JOPTION_SELECT_MAX_LEVELS="- Select Max Levels -"
 JOPTION_SELECT_TAG="- Select Tag -"
+JOPTION_SELECT_CHECKED_OUT="- Select Checked Out By -"
```
Comment (infograf768) [inline administrator/language/en-GB/en-GB.ini]: Please alpha order this string in the JOPTION_SELECT_ group

Comment (nvyush) [reply administrator/language/en-GB/en-GB.ini]: Done

**Comment by bayareajenn** [comment]
I have tested this item :white_check_mark: successfully on 8868a8f581180ce2ae7792c81bf866c57cff1884<br /><br />I've just tested the latest but the item isn't showing it very well. The new labels of All Users and Super User are working fine.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17491">issues.joomla.org/tracker/joomla-cms/17491</a>.</sub>

**Comment by nvyush** [comment]
@infograf768 Could you tell me please, why continuous-integration/drone/pr check is not OK after I removed the found error?

**Comment by nvyush** [comment]
Conflicts with branch 'staging' after merging PR #17668 have been resolved.

**Comment by infograf768** [comment]
Looks OK here.
@mbabker can go or not in 3.8.0 ?

**Comment by nvyush** [comment]
Conflicts with branch 'staging' after merging PR #17748 have been resolved.

**Comment by alikon** [comment]
as a new feature this should be rebased on the 4.0 branch

**Comment by nvyush** [comment]
Does anyone except me need this feature? If yes, I will rebase it on branch 4.0.

**Comment by bayareajenn** [comment]
I don't need it. But I can imagine that on a large site with multiple folks logged in all the time that are editing it might be helpful.

**Comment by Quy** [comment]
@HLeithner Decision please for J4?

**Comment by HLeithner** [comment]
@nvyush it would be great if you can rebase it for j4.

I'm closing this, thx for the work and I hope you can rebase it soon.

**Pull request closed without merging**